File: sandbox/golang/instrument.go
// instrument.go — source-level instrumentation for Go sandbox jobs.
//
// The worker keeps rich array compare/swap traces when it can detect them.
// For algorithms whose important state is not a []int slice, it also records
// lightweight source-line progress so the frontend always has a usable trace.
package main

import (
	"fmt"
	"regexp"
	"strings"
)

var (
	sliceDeclRe   = regexp.MustCompile(`(?:var\s+)?(\w+)\s*(?::=|=)\s*\[\]int\{`)
	paramRe       = regexp.MustCompile(`func\s+\w+\(\s*(\w+)\s+\[\]int`)
	compareOpRe   = regexp.MustCompile(`[<>]=?|==`)
	mainFuncRe    = regexp.MustCompile(`func\s+main\s*\(\s*\)\s*\{`)
	plainAssignRe = regexp.MustCompile(`(^|[^=!<>])=([^=]|$)`)
)

func findSubjectName(code string) string {
	if m := paramRe.FindStringSubmatch(code); m != nil {
		return m[1]
	}
	if m := sliceDeclRe.FindStringSubmatch(code); m != nil {
		return m[1]
	}
	return ""
}

const runtimePrelude = `
type __step__ struct {
	Type    string   ` + "`json:\"type\"`" + `
	Indices []int    ` + "`json:\"indices,omitempty\"`" + `
	Array   []int    ` + "`json:\"array,omitempty\"`" + `
	Line    int      ` + "`json:\"line,omitempty\"`" + `
	Info    string   ` + "`json:\"info,omitempty\"`" + `
}

var __steps__ []__step__
var __line_steps__ []__step__
var __prev__ []int

func __record_line__(line int) {
	if len(__line_steps__) >= 1200 {
		return
	}
	__line_steps__ = append(__line_steps__, __step__{
		Type: "highlight",
		Line: line,
		Info: fmt.Sprintf("executing Go source line %d", line),
	})
}

func __record_compare__(i, j int, arr []int) {
	if len(__steps__) >= 2000 {
		return
	}
	cp := make([]int, len(arr))
	copy(cp, arr)
	if __prev__ == nil {
		__prev__ = append([]int(nil), cp...)
	}
	__steps__ = append(__steps__, __step__{
		Type:    "compare",
		Indices: []int{i, j},
		Array:   cp,
		Info:    fmt.Sprintf("comparing arr[%d] and arr[%d]", i, j),
	})
}

func __record_mutation__(arr []int) {
	if len(__steps__) >= 2000 {
		return
	}
	cp := make([]int, len(arr))
	copy(cp, arr)
	if __prev__ != nil && len(__prev__) == len(cp) {
		changed := []int{}
		for k := range cp {
			if cp[k] != __prev__[k] {
				changed = append(changed, k)
			}
		}
		if len(changed) == 2 {
			__steps__ = append(__steps__, __step__{
				Type:    "swap",
				Indices: changed,
				Array:   cp,
				Info:    fmt.Sprintf("swapping arr[%d] and arr[%d]", changed[0], changed[1]),
			})
		} else if len(changed) == 1 {
			__steps__ = append(__steps__, __step__{
				Type:    "set",
				Indices: changed,
				Array:   cp,
				Info:    fmt.Sprintf("setting arr[%d] = %d", changed[0], cp[changed[0]]),
			})
		}
	}
	__prev__ = cp
}

func __finalize__() {
	if len(__steps__) == 0 && len(__line_steps__) > 0 {
		__steps__ = append(__steps__, __line_steps__...)
	}
	if len(__steps__) == 0 {
		__steps__ = append(__steps__, __step__{Type: "done", Info: "execution completed"})
	} else {
		last := __steps__[len(__steps__)-1]
		__steps__ = append(__steps__, __step__{Type: "done", Array: last.Array, Info: "done"})
	}
	__emit_steps__()
}

func __emit_steps__() {
	b, _ := json.Marshal(__steps__)
	fmt.Println("__STEPS_JSON__" + string(b))
}
`

func shouldTraceLine(trimmed string) bool {
	if trimmed == "" || strings.HasPrefix(trimmed, "//") {
		return false
	}
	if strings.HasPrefix(trimmed, "package ") || strings.HasPrefix(trimmed, "import ") ||
		strings.HasPrefix(trimmed, "func ") || strings.HasPrefix(trimmed, "type ") ||
		strings.HasPrefix(trimmed, "const ") || strings.HasPrefix(trimmed, "var ") {
		return false
	}
	if trimmed == "{" || trimmed == "}" || trimmed == ")" ||
		strings.HasPrefix(trimmed, "} else") || strings.HasPrefix(trimmed, "else ") ||
		strings.HasPrefix(trimmed, "case ") || strings.HasPrefix(trimmed, "default:") {
		return false
	}
	return strings.HasPrefix(trimmed, "if ") || strings.HasPrefix(trimmed, "for ") ||
		strings.HasPrefix(trimmed, "switch ") || strings.HasPrefix(trimmed, "return") ||
		strings.HasPrefix(trimmed, "break") || strings.HasPrefix(trimmed, "continue") ||
		strings.Contains(trimmed, ":=") || plainAssignRe.MatchString(trimmed) ||
		strings.HasSuffix(trimmed, "++") || strings.HasSuffix(trimmed, "--") ||
		strings.Contains(trimmed, "(")
}

func injectLineTracing(code string) string {
	lines := strings.Split(code, "\n")
	out := make([]string, 0, len(lines)*2)
	inImportBlock := false
	for index, line := range lines {
		trimmed := strings.TrimSpace(line)
		if strings.HasPrefix(trimmed, "import (") {
			inImportBlock = true
			out = append(out, line)
			continue
		}
		if inImportBlock {
			out = append(out, line)
			if trimmed == ")" {
				inImportBlock = false
			}
			continue
		}
		if shouldTraceLine(trimmed) {
			indent := line[:len(line)-len(strings.TrimLeft(line, " \t"))]
			out = append(out, fmt.Sprintf("%s__record_line__(%d)", indent, index+1))
		}
		out = append(out, line)
	}
	return strings.Join(out, "\n")
}

func instrument(code string) (string, string) {
	subject := findSubjectName(code)
	code = injectLineTracing(code)
	lines := strings.Split(code, "\n")
	out := make([]string, 0, len(lines)*2)

	var idxRe *regexp.Regexp
	if subject != "" {
		idxRe = regexp.MustCompile(regexp.QuoteMeta(subject) + `\[([^\]]+)\]`)
	}

	for _, line := range lines {
		if subject != "" && compareOpRe.MatchString(line) {
			matches := idxRe.FindAllStringSubmatch(line, -1)
			if len(matches) >= 2 {
				indent := line[:len(line)-len(strings.TrimLeft(line, " \t"))]
				out = append(out, fmt.Sprintf("%s__record_compare__((%s), (%s), %s)", indent, matches[0][1], matches[1][1], subject))
			}
		}

		out = append(out, line)
		if subject != "" && idxRe.MatchString(line) && plainAssignRe.MatchString(line) {
			indent := line[:len(line)-len(strings.TrimLeft(line, " \t"))]
			out = append(out, fmt.Sprintf("%s__record_mutation__(%s)", indent, subject))
		}
	}

	joined := strings.Join(out, "\n")
	joined = mainFuncRe.ReplaceAllString(joined, "func main() {\n\tdefer __finalize__()\n")
	return joined, subject
}

File: sandbox/cpp/instrument.py
"""
instrument.py — source-level instrumentation for C/C++ sandbox jobs.

Same pragmatic strategy documented for the Go and Java sandboxes: inject
step-recording calls at compile time around comparisons and assignments
involving the detected "subject" std::vector<int> (or int[]) array. C/C++
is marked "beta" in the docs/UI per requirement_doc.md §3.4 — instrumentation
fidelity here is the least mature of the five languages.
"""

import re

VECTOR_PARAM_RE = re.compile(r"std::vector<int>\s*&?\s*(\w+)")
VECTOR_DECL_RE = re.compile(r"std::vector<int>\s+(\w+)\s*=\s*\{")
ARRAY_PARAM_RE = re.compile(r"\bint\s+(\w+)\s*\[\s*\]")
COMPARE_OP_RE = re.compile(r"[<>]=?|==")
MAIN_RE = re.compile(r"\bint\s+main\s*\([^)]*\)\s*\{")

RUNTIME_PRELUDE = """
#include <vector>
#include <string>
#include <sstream>
#include <iostream>

namespace __algoviz__ {
    inline std::string vecToJson(const std::vector<int>& arr) {
        std::ostringstream ss;
        ss << "[";
        for (size_t i = 0; i < arr.size(); i++) {
            if (i > 0) ss << ",";
            ss << arr[i];
        }
        ss << "]";
        return ss.str();
    }

    static std::ostringstream __steps__;
    static bool __first__ = true;
    static std::vector<int> __prev__;
    static bool __has_prev__ = false;
    static int __count__ = 0;

    inline void appendStep(const std::string& json) {
        if (__count__ >= 2000) return;
        __count__++;
        if (!__first__) __steps__ << ",";
        __first__ = false;
        __steps__ << json;
    }

    inline void recordCompare(int i, int j, const std::vector<int>& arr) {
        if (!__has_prev__) { __prev__ = arr; __has_prev__ = true; }
        std::ostringstream ss;
        ss << "{\\"type\\":\\"compare\\",\\"indices\\":[" << i << "," << j << "],\\"array\\":"
           << vecToJson(arr) << ",\\"info\\":\\"comparing arr[" << i << "] and arr[" << j << "]\\"}";
        appendStep(ss.str());
    }

    inline void recordMutation(const std::vector<int>& arr) {
        if (__has_prev__ && __prev__.size() == arr.size()) {
            std::vector<int> changed;
            for (size_t k = 0; k < arr.size(); k++) {
                if (__prev__[k] != arr[k]) changed.push_back((int)k);
            }
            if (changed.size() == 2) {
                std::ostringstream ss;
                ss << "{\\"type\\":\\"swap\\",\\"indices\\":[" << changed[0] << "," << changed[1] << "],\\"array\\":"
                   << vecToJson(arr) << ",\\"info\\":\\"swapping arr[" << changed[0] << "] and arr[" << changed[1] << "]\\"}";
                appendStep(ss.str());
            } else if (changed.size() == 1) {
                std::ostringstream ss;
                ss << "{\\"type\\":\\"set\\",\\"indices\\":[" << changed[0] << "],\\"array\\":"
                   << vecToJson(arr) << ",\\"info\\":\\"setting arr[" << changed[0] << "]\\"}";
                appendStep(ss.str());
            }
        }
        __prev__ = arr;
        __has_prev__ = true;
    }

    inline void finalizeSteps() {
        if (__count__ == 0) {
            __steps__ << "{\\"type\\":\\"done\\",\\"info\\":\\"execution completed \\u2014 no array-based state changes detected\\"}";
        } else {
            __steps__ << ",{\\"type\\":\\"done\\",\\"array\\":" << vecToJson(__prev__) << ",\\"info\\":\\"done\\"}";
        }
        std::cout << "__STEPS_JSON__[" << __steps__.str() << "]" << std::endl;
    }

    struct FinalizeGuard {
        ~FinalizeGuard() { finalizeSteps(); }
    };
}
using __algoviz__::recordCompare;
using __algoviz__::recordMutation;
"""


def find_subject_name(code: str) -> str:
    m = VECTOR_PARAM_RE.search(code)
    if m:
        return m.group(1)
    m = VECTOR_DECL_RE.search(code)
    if m:
        return m.group(1)
    m = ARRAY_PARAM_RE.search(code)
    if m:
        return m.group(1)
    return ""


def _find_matching_brace(code: str, open_idx: int) -> int:
    depth = 0
    for i in range(open_idx, len(code)):
        if code[i] == "{":
            depth += 1
        elif code[i] == "}":
            depth -= 1
            if depth == 0:
                return i
    return -1


def instrument(code: str) -> str:
    subject = find_subject_name(code)

    if subject:
        idx_re = re.compile(re.escape(subject) + r"\[([^\]]+)\]")
        assign_re = re.compile(re.escape(subject) + r"\[[^\]]+\]\s*=[^=]")

        lines = code.split("\n")
        out = []
        for line in lines:
            if COMPARE_OP_RE.search(line):
                matches = idx_re.findall(line)
                if len(matches) >= 2:
                    out.append(f"__algoviz__::recordCompare(({matches[0]}), ({matches[1]}), {subject});")
            out.append(line)
            if assign_re.search(line) or ("std::swap" in line and subject + "[" in line):
                out.append(f"__algoviz__::recordMutation({subject});")
        code = "\n".join(out)

    m = MAIN_RE.search(code)
    if m:
        open_brace_idx = code.index("{", m.end() - 1)
        code = (
            code[:open_brace_idx + 1]
            + "\n__algoviz__::FinalizeGuard __algoviz_finalize_guard__;\n"
            + code[open_brace_idx + 1:]
        )

    return RUNTIME_PRELUDE + "\n" + code
